### PR TITLE
Support packet reallocation: forget type for pointers to packet

### LINKED ISSUE #226: support for packet reallocation (when its size changes)?
A question came up about whether the verifier supports packet reallocation (when its size changes).

Opening this issue to track the answer, even if the answer is that it's up to the hooks/helpers to somehow do it rather than the verifier proper.

### PR DESCRIPTION
**Pull request by elazarg**
* Add to function prototypes information about possibility of change to packet size.
* Handle potential changes by invalidating the type of all variables that may point to the packet, and re-initialize packet size and meta size.

Fix #226.

Signed-off-by: Elazar Gershuni <elazarg@gmail.com>

### CODE REVIEW

**Comment by dthaler** [comment]
My opinion is that #226 should not be closed until there is a test

**Comment by elazarg** [comment]
I'm not going to merge this PR without a test.

File: src/crab/ebpf_domain.hpp [hunk 1]
```diff
@@ -1120,6 +1137,22 @@ class ebpf_domain_t final {
...
+        using namespace dsl_syntax;
+
+        inv -= variable_t::packet_size();
+        inv -= variable_t::meta_offset();
+
+        inv += 0 <= variable_t::packet_size();
+        inv += variable_t::packet_size() < MAX_PACKET_OFF;
+        if (global_program_info.type.context_descriptor->meta >= 0) {
+            inv += variable_t::meta_offset() <= 0;
+            inv += variable_t::meta_offset() >= -4098;
```
Comment (dthaler) [inline src/crab/ebpf_domain.hpp:1150]: A comment would be helpful to explain where the number -4098 came from.
(Yes I know the line was just moved from old line 1143 so this is not a comment on this change per se.)

File: src/helpers.hpp [hunk 2]
```diff
@@ -14,6 +14,9 @@ struct EbpfHelperPrototype {
     // Arguments are passed in registers R1 to R5.
     ebpf_argument_type_t argument_type[5];
 
+    // Side effect: can this helper perform packet reallocation
```
Comment (dthaler) [inline src/helpers.hpp:17]:
```suggestion
    // Side effect: can this helper perform packet reallocation.
```

**Review by dthaler — APPROVED** [review]

File: src/crab/ebpf_domain.hpp (hunk 1 above)
Comment (elazarg) [reply src/crab/ebpf_domain.hpp:1150]: I think I made this number up, but I'm not sure.

Comment (elazarg) [reply src/crab/ebpf_domain.hpp:1150]: I think this number should be taken from `ebpf_platform`.

File: src/crab/ebpf_domain.hpp [hunk 3]
```diff
@@ -1136,14 +1169,8 @@ class ebpf_domain_t final {
...
 
-        inv += 0 <= variable_t::packet_size();
-        inv += variable_t::packet_size() < MAX_PACKET_OFF;
-        if (global_program_info.type.context_descriptor->meta >= 0) {
-            inv += variable_t::meta_offset() <= 0;
-            inv += variable_t::meta_offset() >= -4098;
-        } else {
-            inv.assign(variable_t::meta_offset(), 0);
-        }
+        initalize_packet(inv);
```
Comment (dthaler) [inline src/crab/ebpf_domain.hpp:1172]: Looks like a typo in the function name, I think this should be "initialize_packet"

Comment (elazarg) [reply src/crab/ebpf_domain.hpp:1172]: Thanks. I fixed it directly on master.

**Pull request merged**